tests/test_intl/test_intl.py

```diff
@@ -681,50 +687,73 @@ def test_translation_progress_classes_true(app):
     assert len(doctree[0]) == 20
 
 
+@pytest.mark.parametrize('i', range(120))
 @sphinx_intl
 # use individual shared_result directory to avoid "incompatible doctree" error
-@pytest.mark.sphinx(testroot='builder-gettext-dont-rebuild-mo')
-def test_gettext_dont_rebuild_mo(make_app, app_params):
+@pytest.mark.sphinx(testroot='builder-gettext-dont-rebuild-mo', freshenv=True)
+def test_gettext_dont_rebuild_mo(monkeypatch, make_app, app_params, i):
+    import time
+
     # --- don't rebuild by .mo mtime
     def get_update_targets(app_):
         app_.env.find_files(app_.config, app_.builder)
         added, changed, removed = app_.env.get_outdated_files(config_changed=False)
         return added, changed, removed
 
-    args, kwargs = app_params
-
-    # phase1: build document with non-gettext builder and generate mo file in srcdir
-    app0 = make_app('dummy', *args, **kwargs)
-    app0.build()
-    time.sleep(0.01)
-    assert (app0.srcdir / 'xx' / 'LC_MESSAGES' / 'bom.mo').exists()
-    # Since it is after the build, the number of documents to be updated is 0
-    update_targets = get_update_targets(app0)
-    assert update_targets[1] == set(), update_targets
-    # When rewriting the timestamp of mo file, the number of documents to be
-    # updated will be changed.
-    mtime = (app0.srcdir / 'xx' / 'LC_MESSAGES' / 'bom.mo').stat().st_mtime
-    os.utime(app0.srcdir / 'xx' / 'LC_MESSAGES' / 'bom.mo', (mtime + 5, mtime + 5))
-    update_targets = get_update_targets(app0)
-    assert update_targets[1] == {'bom'}, update_targets
-
-    # Because doctree for gettext builder can not be shared with other builders,
-    # erase doctreedir before gettext build.
-    shutil.rmtree(app0.doctreedir)
-
-    # phase2: build document with gettext builder.
-    # The mo file in the srcdir directory is retained.
-    app = make_app('gettext', *args, **kwargs)
-    app.build()
-    time.sleep(0.01)
-    # Since it is after the build, the number of documents to be updated is 0
-    update_targets = get_update_targets(app)
-    assert update_targets[1] == set(), update_targets
-    # Even if the timestamp of the mo file is updated, the number of documents
-    # to be updated is 0. gettext builder does not rebuild because of mo update.
-    os.utime(app0.srcdir / 'xx' / 'LC_MESSAGES' / 'bom.mo', (mtime + 10, mtime + 10))
-    update_targets = get_update_targets(app)
-    assert update_targets[1] == set(), update_targets
+    with monkeypatch.context() as m:
+        if os.name != 'posix':
+            offset = time.time_ns()
+            t0 = time.perf_counter_ns()
+            monkeypatch.setattr(time, 'time_ns', lambda: offset + time.perf_counter_ns() - t0)
```

Comment:
> Actually the whole issue stems from the fact that time.time_ns() has a resolution of [318 microseconds on Windows](https://peps.python.org/pep-0564/#windows) vs [84ns on Unix-based systems](https://peps.python.org/pep-0564/#linux). That's likely the reason why my two files end up having the same st_mtime_ns value but not on Unix (since they are effectively created at different points in time). If Windows had a good time.time_ns() resolution, then this issue would likely disappear.

During this monkeypatching, does `time.perf_counter_ns` act as the improved-resolution timer that you mention?

And if so: why is it necessary to _also_ increase the interval size for the modified-timestamps (from `+ 5` to `+ 1000` for example)?

Reply:
I'm trying something now so don't worry about the state of that PR. I'm doing a totally different approach and trying to see what really happens actually.